### ISSUE TITLE
iCustom Parameters

### ISSUE THREAD

**Issue by BinkyBonkers** [vdemydiuk/mtapi#70]
What is the right way to call iCustom with different parameter types, i.e. a mixture of ints, doubles and strings?  Please can you give me an example.  Many thanks.

**Comment by mhdbzr** [comment]
I think it's good to pass the value array and typeOfValues array to metatrader to decide to call with iCustom function (depend on number of inputs*) and handle parameters type.

* In this case, MtApi supports 4-100 input parameters with iCustom_4 - iCustom_100 functions!

**Comment by vdemydiuk** [comment]
It's no problem to pass values with same type to MetaTrader. But when we talk about different types then we have restriction of MQL: it does not allow to cast to values types in runtime (maybe I don't know how it can be do it).

double  iCustom(
   string       symbol,           // symbol
   int          timeframe,        // timeframe
   string       name,             // path/name of the custom indicator compiled program
   ...                            // custom indicator input parameters (if necessary)
   int          mode,             // line index
   int          shift             // shift
   );

I will be glad if you or somebody else will find the way how to prepare list of custom input parameters in runtime.

**Comment by mhdbzr** [comment]
Yes, you are right. I was thinking metatrader handles inline if as a runtime casting, but when I try calling iCustom with string and double inputs with inline if, I failed :( .
If I find any way, I will update this topic.

**Comment by sej2016** [comment]
It is not possible to call ICustom directly with the parameters in string format separed by comma according to the MQL4 documentation.
ICustom(symbol ,timeframe ,name , param 1, param 2, param 3,....., mode, shift). 
I dont know if this match, it is only a ideas.

**Comment by vdemydiuk** [comment]
I think for MT5 we can try to use function MQL IndicatorCreate instead iCustom:
https://www.mql5.com/en/docs/series/indicatorcreate

Function IndicatorCreate uses array of parameters of MqlParam type.
MqlParam can have different type of values (long, double, string).
I will implement functions IndicatorCreate, IndicatorRelease with related issue #92.

For MT4... I still do not know.

**Comment by ariyanto7** [comment]
> I think for MT5 we can try to use function MQL IndicatorCreate instead iCustom:
> https://www.mql5.com/en/docs/series/indicatorcreate
> 
> Function IndicatorCreate uses array of parameters of MqlParam type.
> MqlParam can have different type of values (long, double, string).
> I will implement functions IndicatorCreate, IndicatorRelease with related issue #92.
> 
> For MT4... I still do not know.

mt4 we use iCustom. almost the same i think.

> It is not possible to call ICustom directly with the parameters in string format separed by comma according to the MQL4 documentation.
> ICustom(symbol ,timeframe ,name , param 1, param 2, param 3,....., mode, shift).
> I dont know if this match, it is only a ideas.

if the indicator has 10 params, and we only need 3 params to be change, do we need to put all 10 params in the iCustom?

**Comment by afsalme** [comment]
I ended up here in this git looking for the same to achieve calling ICustom directly with the parameters in string format separated by comma.
And see none here also have a solution for this.
But in further searching I found someone have done this. Check below link
https://www.mql5.com/en/blogs/post/721838
Search for 'SIGNALStrategyInputs' in above page.
Anyone able to decode the logic how this is done?

**Comment by mhdbzr** [comment]
I see this in that blog post:
`SIGNALStrategyInputs = 1,2.0, C'0 / 128/0 ', " text ", D'2018.11.08', false`
I remember a project, that I was trying to dump all indicator values on the chart.
my solution and this raw value replacement are similar! I was creating and compiling new indicator to ask me referenced indicator name and its buffer index, then it return buffer value. And I was writing iCustom function and related inputs to compile and work for me as a trick !
I did not test Afsalme linked product, but it's copied settings are similar to what I did before.

**Comment by afsalme** [comment]
> I see this in that blog post:
> `SIGNALStrategyInputs = 1,2.0, C'0 / 128/0 ', " text ", D'2018.11.08', false`
> I remember a project, that I was trying to dump all indicator values on the chart.
> my solution and this raw value replacement are similar! I was creating and compiling new indicator to ask me referenced indicator name and its buffer index, then it return buffer value. And I was writing iCustom function and related inputs to compile and work for me as a trick !
> I did not test Afsalme linked product, but it's copied settings are similar to what I did before.

I am also trying to do the same. Can you help me in this.
I have attached the indicator mq4 code. Please be kind to check it.
[IndicatorAlerts.txt](https://github.com/vdemydiuk/mtapi/files/4530937/IndicatorAlerts.txt)

**Comment by mhdbzr** [comment]
I see you're using `IndicatorParameters` from https://www.mql5.com/en/docs/series/indicatorparameters . it is MQL5 not implemented in MT4.

What I did, briefly: my EA writes a new indicator/EA source code in a file and then compile & open it to read result by calling its buffer or exported function.

**Comment by afsalme** [comment]
> I see you're using `IndicatorParameters` from https://www.mql5.com/en/docs/series/indicatorparameters . it is MQL5 not implemented in MT4.
> 
> What I did, briefly: my EA writes a new indicator/EA source code in a file and then compile & open it to read result by calling its buffer or exported function.

Very interesting Solution you have mhdbzr. 🥇 💯 
Thank you for sharing your idea. I will try myself to implement your way of achieving this. 
It will be great if you can share your code, if you don't mind. :)

**Comment by mhdbzr** [comment]
I was trying to share it, but it has too many other parts and it was not possible to share it.
Note: if you re-write imported Indicator in run-time with the same name, mt4 shutdown by crash, probably.
so, it's better to use new name for new Indicator, if you're watching all attached indicators on the chart and re-building the Indicator each minute or frequently by chart changes.
